### PR TITLE
fix(auth): redirect to general login if user is in invited state for single_org_mode

### PR DESCRIPTION
**Pull request by JoshFerge**
- If a user is kicked out of an organization (from say admin enabling 2fa), and they have a page bookmarked that is an organization specific view, they will continually be sent back to the org login ad infinitum after attempting to log in.
- this PR makes it so when logging in via the org page, if the user is in this invited state (no user on OrgMember), they will instead be re-directed to the general login page, which for single org mode, will take them to the "No Organizations" page.
- the mechanism of this is the original url being set in the `_next` session variable.
- Only enabled on Single Org Mode as to not interfere with multple-org-account logins (see https://github.com/getsentry/sentry/issues/2347) -- i'm not sure how important it is for us to maintain this but doing so for now.

TODO:

- [x] fix code so existing tests pass
- [x] write test for this case
- [x] think if any edge cases would make this solution invalid
- [x] shouldn't mess with SSO jit provisioning as logic is added in basic_auth, which sso doesn't hit I believe.

### CODE REVIEW

**Review by leedongwei — COMMENTED** [review]
First pass looks good to me. I've verified that it stops the redirect loop.

We're not writing anything so let's deploy to `staging` to check.

File: tests/sentry/web/frontend/test_react_page.py [hunk 1]
```diff
@@ -72,3 +72,22 @@ def test_inactive_superuser_bypasses_server_auth(self):
...
+        member = OrganizationMember.objects.get(organization=self.organization, user=user)
+        member.email = "foo@example.com"
+        member.user = None
+        member.save()
+
+        self.login_as(user)
+        path = reverse("sentry-organization-home", args=[self.organization.slug])
+        resp = self.client.get(path)
+        assert resp.status_code == 302
+        # assert resp["Location"] != reverse("sentry-auth-organization", args=[org.slug])
```
Comment (EvanPurkhiser) [inline tests/sentry/web/frontend/test_react_page.py:89]: Was this not working?

Comment (JoshFerge) [reply tests/sentry/web/frontend/test_react_page.py]: just fixed!

**Review by evanpurkhiser — APPROVED** [review]

**Comment by JoshFerge** [comment]
longer term we probably want to craft a specific page for this. "You have received an invite to the org, but please check your email to accept the invite". -> this is assuming that we don't want to auto-accept an invite if the user has a pending one and they try to login to the org.

**Comment by JoshFerge** [comment]
Need to rethink this a bit, closing for now.

File: tests/sentry/web/frontend/test_auth_organization_login.py [hunk 2]
```diff
@@ -652,3 +654,71 @@ def test_flow_as_unauthenticated_existing_user_legacy_identity_migration(self):
...
+    def test_basic_auth_flow_as_invited_user(self):
+
+        user = self.create_user("foor@example.com")
+        self.create_member(organization=self.organization, user=user)
+        member = OrganizationMember.objects.get(organization=self.organization, user=user)
+        member.email = "foor@example.com"
+        member.user = None
+        member.save()
+
+        # session = self.client.session
```
Comment (JoshFerge) [inline tests/sentry/web/frontend/test_auth_organization_login.py:669]: i can't get the session value setting to work here but i'd like to test when `_next` is an OrgView.

File: tests/sentry/web/frontend/test_auth_organization_login.py [hunk 3]
```diff
@@ -652,3 +654,70 @@ def test_flow_as_unauthenticated_existing_user_legacy_identity_migration(self):
         # Ensure the ident was migrated from the legacy identity
         updated_ident = AuthIdentity.objects.get(id=user_ident.id)
         assert updated_ident.ident == "foo@new-domain.com"
+
+    @override_settings(SENTRY_SINGLE_ORGANIZATION=True)
+    @with_feature({"organizations:create": False})
+    def test_basic_auth_flow_as_invited_user(self):
```
Comment (JoshFerge) [inline tests/sentry/web/frontend/test_auth_organization_login.py:660]: this is the money test, as before, with this value in `_next`, we'd successfully auth them and try to redirect back to the original org page they tried to go to which would then take them back here.

File: src/sentry/web/frontend/auth_login.py [hunk 4]
```diff
@@ -193,6 +193,19 @@ def handle_basic_auth(self, request, **kwargs):
 
                 if not user.is_active:
                     return self.redirect(reverse("sentry-reactivate-account"))
+                if organization and settings.SENTRY_SINGLE_ORGANIZATION:
+                    try:
+                        om = OrganizationMember.objects.get(
+                            organization=organization, email=user.email
```
Comment (JoshFerge) [inline src/sentry/web/frontend/auth_login.py:199]: is it ever possible for a user not to have an email and have an org invite?

**Review by leedongwei — APPROVED** [review]
LGTM
